<a href="https://colab.research.google.com/github/jesschannn/datasci_2_manipulation/blob/main/datasci_2_manipulationreal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning with Pandas

In [26]:
import pandas as pd
import numpy as np

In [27]:
df = pd.read_csv('healthcare_data_cleaning.csv')
df

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

## Changing "Missing" to "NaN"

In [28]:
df.replace('missing', pd.np.nan, inplace=True)
df

<ipython-input-28-31a068059994>:1: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead.
  df.replace('missing', pd.np.nan, inplace=True)


Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59        NaN            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996                NaN        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195                   NaN   
2        315.0709305262176                   NaN                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

## Removing Duplicates

In [29]:
df.drop_duplicates(inplace=True)
df

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104994          53  Female      Michelleburgh       South Dakota   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998         NaN  Female       Heidiborough                NaN   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes                NaN           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104994            No                 No           Card     Dr. Williams   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104994               Flu                 Med_B           Emergency   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59        NaN            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104994                 68        109            172             75   
104995                 65        112            169             61   
104996                NaN        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195                   NaN   
2        315.0709305262176                   NaN                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104994   488.0316041186948                   167                    25   
104995  309.164

## Removed White Space and Special Characters

In [30]:
import pandas as pd
import re
import numpy as np

In [31]:
# Function to remove white space and special characters from a value
def clean_column_names(df1):
    # Define a helper function to clean column names
    def clean_name(name):
        cleaned_name = re.sub(r'[^a-zA-Z0-9]', '', name)
        return cleaned_name.lower()

    # Rename columns using the helper function
    # This is using a list comprehend - e.g., we have a list to the right of the equals sign,
    # and inside the list, we are applying our function, for every col (or X) that exists in df.columns
    df1.columns = [clean_name(col) for col in df1.columns]
    return df

# Apply the clean_value function to all columns
df1 = clean_column_names(df)

df1

patientage  gender    cityofresidence stateofresidence hasinsurance  \
0              45   Other         West Brian         Kentucky          Yes   
1              25  Female   East Jocelynfurt         Colorado           No   
2              51   Other  South Lindseyland             Ohio          Yes   
3              18  Female         Taylorfort     Pennsylvania          Yes   
4              48  Female          Birdmouth          Montana           No   
...           ...     ...                ...              ...          ...   
104994         53  Female      Michelleburgh     South Dakota           No   
104995         74   Other       West Anthony         Oklahoma          Yes   
104996         37    Male          Mckeetown             Utah          Yes   
104997         51  Female   Lake Lisachester         Nebraska          Yes   
104998        NaN  Female       Heidiborough              NaN          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasediagnosed  \
0                   Yes     Insurance    Dr. Williams              Flu   
1                   Yes          Card     Dr. Johnson              Flu   
2                    No          Cash    Dr. Williams         Covid-19   
3                   NaN          Cash    Dr. Williams              Flu   
4                    No     Insurance       Dr. Brown             None   
...                 ...           ...             ...              ...   
104994               No          Card    Dr. Williams              Flu   
104995               No          Card    Dr. Williams         Covid-19   
104996               No          Cash       Dr. Jones         Covid-19   
104997              Yes     Insurance    Dr. Williams          Allergy   
104998               No          Cash       Dr. Smith         Covid-19   

       medicationprescribed typeofappointment averageheartrate averagebp  \
0                     Med_E           General               66       111   
1                     Med_B        Specialist               59       NaN   
2                     Med_D           General               79       119   
3                     Med_A           General               99       115   
4                     Med_C        Specialist               93        89   
...                     ...               ...              ...       ...   
104994                Med_B         Emergency               68       109   
104995                Med_E        Specialist               65       112   
104996                Med_C           General              NaN       111   
104997                Med_D         Follow-Up               89       118   
104998                Med_A         Follow-Up               74       123   

       heightincm weightinkg          paymentdue lastvisitdaysago  \
0             183         70   65.45721578126224              193   
1             174         58  430.68365678679174              195   
2             161         56   315.0709305262176              NaN   
3             171         52   320.2998987723972              175   
4             186         87   264.2147372473319              188   
...           ...        ...                 ...              ...   
104994        172         75   488.0316041186948              167   
104995        169         61  309.16444606680665              183   
104996        165         67                 NaN              191   
104997        179         71   320.8304180033907              172   
104998        167         75   5.341231420036174              180   

       visitdurationmins numberoftests    prescriptioncost  
0                     92             5   15.71751735532889  
1                    NaN             4   80.78647284463952  
2                     37             2    64.3921393009105  
3                    108             2   8.872859215315316  
4                    113             3   77.48311304645532  
...                  ...           ...                 ...  
104994      

## Handling Missing Values

In [32]:
import re
import numpy as np

In [33]:
df['paymentdue'] = pd.to_numeric(df['paymentdue'], errors='coerce')
df['prescriptioncost'] = pd.to_numeric(df['prescriptioncost'], errors='coerce')

In [34]:
df['patientage'] = pd.to_numeric(df['patientage'], errors='coerce')
df['averageheartrate'] = pd.to_numeric(df['averageheartrate'], errors='coerce')
df['averagebp'] = pd.to_numeric(df['averagebp'], errors='coerce')
df['weightinkg'] = pd.to_numeric(df['weightinkg'], errors='coerce')
df['heightincm'] = pd.to_numeric(df['heightincm'], errors ='coerce')

In [35]:
df['lastvisitdaysago'] = pd.to_numeric(df['lastvisitdaysago'], errors='coerce')
df['visitdurationmins'] = pd.to_numeric(df['visitdurationmins'], errors='coerce')
df['numberoftests'] = pd.to_numeric(df['numberoftests'], errors = 'coerce')

In [36]:
df1=df.dropna()
df1

patientage  gender   cityofresidence stateofresidence hasinsurance  \
0             45.0   Other        West Brian         Kentucky          Yes   
4             48.0  Female         Birdmouth          Montana           No   
7             52.0    Male      Vanessaburgh      Connecticut          Yes   
8             20.0  Female  West Briannabury          Montana           No   
11            40.0  Female        Jonesmouth     Rhode Island           No   
...            ...     ...               ...              ...          ...   
104992        44.0  Female    Port Tracistad           Alaska          Yes   
104993        42.0  Female     West Annaport          Florida          Yes   
104994        53.0  Female     Michelleburgh     South Dakota           No   
104995        74.0   Other      West Anthony         Oklahoma          Yes   
104997        51.0  Female  Lake Lisachester         Nebraska          Yes   

       visitedlastmonth paymentmethod preferreddoctor diseasediagnosed  \
0                   Yes     Insurance    Dr. Williams              Flu   
4                    No     Insurance       Dr. Brown             None   
7                   Yes     Insurance    Dr. Williams              Flu   
8                   Yes          Card       Dr. Jones              Flu   
11                   No          Card       Dr. Jones         Covid-19   
...                 ...           ...             ...              ...   
104992               No     Insurance       Dr. Brown             Cold   
104993               No          Cash     Dr. Johnson         Covid-19   
104994               No          Card    Dr. Williams              Flu   
104995               No          Card    Dr. Williams         Covid-19   
104997              Yes     Insurance    Dr. Williams          Allergy   

       medicationprescribed typeofappointment  averageheartrate  averagebp  \
0                     Med_E           General              66.0      111.0   
4                     Med_C        Specialist              93.0       89.0   
7                     Med_E           General              99.0      129.0   
8                     Med_C        Specialist              67.0      108.0   
11                    Med_E        Specialist              65.0      106.0   
...                     ...               ...               ...        ...   
104992                Med_A         Follow-Up              76.0       93.0   
104993                Med_D        Specialist              80.0      129.0   
104994                Med_B         Emergency              68.0      109.0   
104995                Med_E        Specialist              65.0      112.0   
104997                Med_D         Follow-Up              89.0      118.0   

        heightincm  weightinkg  paymentdue  lastvisitdaysago  \
0            183.0        70.0   65.457216             193.0   
4            186.0        87.0  264.214737             188.0   
7            170.0        47.0  387.085044             179.0   
8            173.0        86.0  172.548343             174.0   
11           166.0        82.0  299.142676             193.0   
...            ...         ...         ...               ...   
104992       178.0        66.0  462.565671             189.0   
104993       171.0        78.0  131.356326             185.0   
104994       172.0        75.0  488.031604             167.0   
104995       169.0        61.0  309.164446             183.0   
104997       179.0        71.0  320.830418             172.0   

        visitdurationmins  numberoftests  prescriptioncost  
0                    92.0            5.0         15.717517  
4                   113.0            3.0         77.483113  
7                   108.0            2.0         76.965926  
8                    91.0            2.0        142.317108  
11                   92.0            1.0         77.715360  
...                   ...            ...               ...  
104992               55.0            3.0         

In [37]:
missing_values = df.isnull().sum()
missing_values

patientage              5000
gender                  5000
cityofresidence         5000
stateofresidence        5000
hasinsurance            5000
visitedlastmonth        5000
paymentmethod           5000
preferreddoctor         5000
diseasediagnosed        5000
medicationprescribed    5000
typeofappointment       5000
averageheartrate        5000
averagebp               5000
heightincm              5000
weightinkg              5000
paymentdue              5000
lastvisitdaysago        5000
visitdurationmins       5000
numberoftests           5000
prescriptioncost        5000
dtype: int64

## Summary of Changes Made to the Dataset

After loading the dataset into Google Colab, I changed all of the "missing" data into NaN. Then I removed the duplicates. Removing the duplicates brought the amount of records / rows down from 10,500 to 10,000. Then I removed white space and special characters from the dataset. After that, I identified and handled missing values in the dataset. I removed all of the rows that had missing data. Lastly, I cleaned the column names. In the original dataset, there were white spaces in the column names. There are now underscore characters where the white spaces used to be.

# Data Transformation

## Creating / Renaming Columns

In [38]:
df1.rename(columns = {'patientage':'age', 'cityofresidence': 'city', 'stateofresidence': 'state', 'diseasediagnosed':'diagnosis', 'medicationprescribed': 'prescription', 'heightinkg': 'height_cm', 'weightinkg':'weight_kg'}, inplace = True)
df1

<ipython-input-38-04258c1b5236>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1.rename(columns = {'patientage':'age', 'cityofresidence': 'city', 'stateofresidence': 'state', 'diseasediagnosed':'diagnosis', 'medicationprescribed': 'prescription', 'heightinkg': 'height_cm', 'weightinkg':'weight_kg'}, inplace = True)


age  gender              city         state hasinsurance  \
0       45.0   Other        West Brian      Kentucky          Yes   
4       48.0  Female         Birdmouth       Montana           No   
7       52.0    Male      Vanessaburgh   Connecticut          Yes   
8       20.0  Female  West Briannabury       Montana           No   
11      40.0  Female        Jonesmouth  Rhode Island           No   
...      ...     ...               ...           ...          ...   
104992  44.0  Female    Port Tracistad        Alaska          Yes   
104993  42.0  Female     West Annaport       Florida          Yes   
104994  53.0  Female     Michelleburgh  South Dakota           No   
104995  74.0   Other      West Anthony      Oklahoma          Yes   
104997  51.0  Female  Lake Lisachester      Nebraska          Yes   

       visitedlastmonth paymentmethod preferreddoctor diagnosis prescription  \
0                   Yes     Insurance    Dr. Williams       Flu        Med_E   
4                    No     Insurance       Dr. Brown      None        Med_C   
7                   Yes     Insurance    Dr. Williams       Flu        Med_E   
8                   Yes          Card       Dr. Jones       Flu        Med_C   
11                   No          Card       Dr. Jones  Covid-19        Med_E   
...                 ...           ...             ...       ...          ...   
104992               No     Insurance       Dr. Brown      Cold        Med_A   
104993               No          Cash     Dr. Johnson  Covid-19        Med_D   
104994               No          Card    Dr. Williams       Flu        Med_B   
104995               No          Card    Dr. Williams  Covid-19        Med_E   
104997              Yes     Insurance    Dr. Williams   Allergy        Med_D   

       typeofappointment  averageheartrate  averagebp  heightincm  weight_kg  \
0                General              66.0      111.0       183.0       70.0   
4             Specialist              93.0       89.0       186.0       87.0   
7                General              99.0      129.0       170.0       47.0   
8             Specialist              67.0      108.0       173.0       86.0   
11            Specialist              65.0      106.0       166.0       82.0   
...                  ...               ...        ...         ...        ...   
104992         Follow-Up              76.0       93.0       178.0       66.0   
104993        Specialist              80.0      129.0       171.0       78.0   
104994         Emergency              68.0      109.0       172.0       75.0   
104995        Specialist              65.0      112.0       169.0       61.0   
104997         Follow-Up              89.0      118.0       179.0       71.0   

        paymentdue  lastvisitdaysago  visitdurationmins  numberoftests  \
0        65.457216             193.0               92.0            5.0   
4       264.214737             188.0              113.0            3.0   
7       387.085044             179.0              108.0            2.0   
8       172.548343             174.0               91.0            2.0   
11      299.142676             193.0               92.0            1.0   
...            ...               ...                ...            ...   
104992  462.565671             189.0               55.0            3.0   
104993  131.356326             185.0               22.0            6.0   
104994  488.031604             167.0               25.0            5.0   
104995  309.164446             183.0               10.0            3.0   
104997  320.830418             172.0               39.0            3.0   

        prescriptioncost  
0              15.717517  
4              77.483113  
7              76.965926  
8             142.317108  
11             77.715360  
...                  ...  
104992         59.746318  
104993        104.430167  
104994         69.542286  
104995         63.204788  
104997         35.622420  

[35828 rows x 20 columns]

## Groupby and Summary Statistics

In [39]:
df1['diagnosis'].value_counts()

Cold        7228
Flu         7185
None        7162
Allergy     7142
Covid-19    7111
Name: diagnosis, dtype: int64

In [40]:
grouped_data = df1.groupby('diagnosis').agg({
    'paymentdue': 'mean'
})
grouped_data

paymentdue
diagnosis            
Allergy    244.926193
Cold       248.925305
Covid-19   251.024933
Flu        248.976765
None       250.975049

## Pivot Tables

In [41]:
pivot_data1 = df1.pivot_table(index='state', columns='diagnosis', values='paymentdue', aggfunc='mean')
pivot_data1

diagnosis          Allergy        Cold    Covid-19         Flu        None
state                                                                     
Alabama         240.960786  248.609521  226.101461  251.667757  247.017729
Alaska          246.874414  251.502446  263.664272  258.038319  243.456534
Arizona         239.957722  237.236259  269.909130  239.960491  258.689307
Arkansas        239.354558  259.014617  261.540967  238.783117  232.068658
California      225.557059  251.017062  235.785484  264.382758  228.873362
Colorado        240.164288  258.180106  250.533892  240.250736  238.313261
Connecticut     245.811139  239.728018  260.657838  252.730255  275.573350
Delaware        260.665090  239.882193  269.667932  265.985774  253.998159
Florida         239.928960  245.346217  247.348811  246.702781  229.165361
Georgia         221.306165  256.352908  255.095251  234.773001  265.681792
Hawaii          236.692683  270.911307  255.748094  252.841528  242.855083
Idaho           253.854796  263.119882  260.646862  258.135070  253.225948
Illinois        254.661234  237.229270  243.922638  239.330714  245.497301
Indiana         254.496929  235.489111  252.047900  243.219246  241.962244
Iowa            259.945412  252.360813  246.594468  245.051934  257.736095
Kansas          251.021902  270.364032  260.165582  246.667749  254.972076
Kentucky        237.786614  245.212134  241.274159  258.348994  255.856212
Louisiana       257.844426  259.866621  238.504555  229.964532  259.517390
Maine           239.137384  254.039079  251.168073  234.403480  240.985222
Maryland        243.539474  228.306232  267.031187  248.545129  246.830835
Massachusetts   249.718465  254.206045  268.612925  240.767372  243.928586
Michigan        243.963660  253.096593  241.262940  264.605927  251.474964
Minnesota       254.464390  246.268391  245.687350  240.064314  220.756805
Mississippi     245.259832  246.220401  237.512259  253.067315  282.289676
Missouri        238.285229  260.849990  239.699137  241.181143  235.979380
Montana         237.558457  237.796736  229.047926  251.171633  247.484898
Nebraska        249.441611  255.979051  243.777130  238.797482  241.868634
Nevada          256.446716  246.066904  264.690035  234.983915  234.865612
New Hampshire   252.060086  248.916536  253.589225  249.462168  252.295211
New Jersey      258.119844  245.346323  237.017478  259.237194  237.855204
New Mexico      234.270483  257.751093  261.719156  274.960727  271.417949
New York        234.319389  229.968658  250.074853  235.805709  273.546601
North Carolina  277.715742  231.462109  256.389154  250.740794  239.615132
North Dakota    239.420286  240.111267  252.011064  250.699877  239.396817
Ohio            230.892101  255.888991  252.778652  252.361582  250.038351
Oklahoma        239.156024  241.735997  255.538500  275.846598  241.522840
Oregon          246.837710  244.560120  245.738695  241.303524  253.953609
Pennsylvania    244.492734  251.177709  244.123365  247.160423  250.104553
Rhode Island    252.766947  248.426074  240.069712  255.868597  266.956769
South Carolina  241.763437  226.199076  260.399702  262.241760  271.332168
South Dakota    246.275292  224.798039  259.715226  246.203961  269.269524
Tennessee       250.654568  259.724616  244.095772  244.417355  245.951320
Texas           248.062019  238.331694  225.972480  245.453806  250.349297
Utah            231.679189  252.091554  261.052211  246.351940  283.695423
Vermont         256.888031  253.597798  249.267140  229.733313  260.767834
Virginia        234.400061  268.208227  246.299345  242.340890  243.135851
Washington      249.485840  263.099751  255.447444  255.907799  272.861506
West Virginia   253.759420  256.873476  241.605478  251.321946  243.283841
Wisconsin       219.841668  241.020430  277.705316  264.111746  240.178808
Wyoming         240.418362  258.835639  262.884809  247.502991  245.108902

In [42]:
pivot_data2 = df1.pivot_table(index='prescription', columns='paymentmethod', values='prescriptioncost', aggfunc='mean')
pivot_data2

paymentmethod       Card       Cash  Insurance
prescription                                  
Med_A          73.471973  75.994445  75.855721
Med_B          75.208148  74.960271  74.756141
Med_C          75.771138  76.048944  75.139967
Med_D          73.942290  75.297394  74.939031
Med_E          74.690713  75.606087  74.073754

In [43]:
pivot_data3 = df1.pivot_table(index='preferreddoctor', columns='typeofappointment', values='lastvisitdaysago', aggfunc='median')
pivot_data3

typeofappointment  Emergency  Follow-Up  General  Specialist
preferreddoctor                                             
Dr. Brown              182.0      183.0    182.0       182.0
Dr. Johnson            182.0      183.0    182.0       182.0
Dr. Jones              183.0      182.0    183.0       183.0
Dr. Smith              183.0      182.0    183.0       183.0
Dr. Williams           182.0      183.0    183.0       182.0

# Introduction to Alternative Libraries

##Polars

In [44]:
pip install polars

In [45]:
import polars as pl

df1 = pl.read_csv("healthcare_data_cleaning.csv")

print("Number of rows:", df1.shape[0])
print("Number of columns:", df1.shape[1])


Number of rows: 105000
Number of columns: 20


In [46]:
import time

In [47]:
dataset_path = 'healthcare_data_cleaning.csv'

In [48]:
start_time = time.time()
polars_df = pd.read_csv(dataset_path)
polars_filtered = polars_df[polars_df['Gender'] == 'Female']
polars_average_age = polars_filtered['Average Heart Rate'].sum()
polars_time = time.time() - start_time

## Load Times

In [49]:
print(f"Polars Execution Time: {polars_time:.4f} seconds")

Polars Execution Time: 0.5654 seconds


### Load Times Observation

I noticed that the execution time for Polars was 0.5654. I initially included Modin in my code and noticed that Modin was slower than Polar. I had to remove Modin from my code because it wouldn't work with the code that I used to change "missing" values into NaN values.